### PR TITLE
gh-104371: Fix calls to __release_buffer__ while an exception is active

### PR DESCRIPTION
**Pull request by JelleZijlstra**
* Issue: gh-104371

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -9157,6 +9158,22 @@ releasebuffer_call_python(PyObject *self, Py_buffer *buffer)
...
 
+static void
+releasebuffer_call_python(PyObject *self, Py_buffer *buffer)
+{
+    // bf_releasebuffer may be called while an exception is already active.
+    // We have no way to report additional errors up the stack, because
+    // this slot returns void, so we simply stash away the active exception
+    // and restore it after the call to Python returns.
+    PyObject *type, *value, *traceback;
+    PyErr_Fetch(&type, &value, &traceback);
```
Comment (kumaraditya303) [inline Objects/typeobject.c]: Use newer `PyErr_GetHandledException`

File: Objects/typeobject.c [hunk 2]
```diff
@@ -9157,6 +9158,22 @@ releasebuffer_call_python(PyObject *self, Py_buffer *buffer)
     Py_DECREF(mv);
 }
 
+static void
+releasebuffer_call_python(PyObject *self, Py_buffer *buffer)
```
Comment (kumaraditya303) [inline Objects/typeobject.c]: Can you inline this in `releasebuffer_call_python_inner` itself? I am not a fan of such once used small functions.

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]

File: Objects/typeobject.c (hunk 1 above)
Comment (JelleZijlstra) [reply Objects/typeobject.c]: I tried that at first, but it didn't work. I think `PyErr_GetHandledException` is meant to be used only when an exception has been caught (you're in the equivalent of an `except` block), while this code is invoked when an exception was just raised.

File: Objects/typeobject.c (hunk 2 above)
Comment (JelleZijlstra) [reply Objects/typeobject.c]: Sure. I didn't do that so I wouldn't have to change the return statements, but I can just add a goto error.

**Comment by JelleZijlstra** [comment]
Thanks for the review @kumaraditya303!

I have made the requested changes; please review again.

File: Objects/typeobject.c (hunk 1 above)
Comment (kumaraditya303) [reply Objects/typeobject.c]: I meant to suggest `PyErr_GetRaisedException`, maybe I made copy paste mistake yesterday. I have pushed change to use it instead.

Comment (JelleZijlstra) [reply Objects/typeobject.c]: Ah thanks, looks like that one is new in 3.12 and I was looking at the 3.11 docs when I wrote this. Thanks for pushing the fix!

**Review by kumaraditya303 — APPROVED** [review]

File: Objects/typeobject.c [hunk 3]
```diff
@@ -9155,6 +9161,10 @@ releasebuffer_call_python(PyObject *self, Py_buffer *buffer)
         PyObject_CallMethodNoArgs(mv, &_Py_ID(release));
```
Comment (markshannon) [inline Objects/typeobject.c:9161]: The return value of this call needs to be checked and disposed of.

File: Objects/typeobject.c [hunk 4]
```diff
@@ -9155,6 +9161,10 @@ releasebuffer_call_python(PyObject *self, Py_buffer *buffer)
...
     }
     Py_DECREF(mv);
+end:
+    assert(!PyErr_Occurred());
```
Comment (markshannon) [inline Objects/typeobject.c:9165]: This needs to checked, not asserted. `PyObject_CallMethodNoArgs(mv, &_Py_ID(release));` above can raise an exception.

**Review by markshannon — COMMENTED** [review]
This fix doesn't appear to be correct.

File: Objects/typeobject.c (hunk 4 above)
Comment (kumaraditya303) [reply Objects/typeobject.c:9165]: Good catch, I missed it, https://github.com/python/cpython/pull/104417 addresses it

**Pull request merged**